monitoring/modules.go

```diff
@@ -0,0 +1,209 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/gravitational/satellite/agent/health"
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// NewModuleChecker creates a new kernel module checker
+func NewModuleChecker(names ...string) ModuleChecker {
+	return ModuleChecker{
+		Modules:      names,
+		moduleGetter: moduleGetterFunc(LoadModules),
+	}
+}
+
+// Name returns name of the checker
+func (r ModuleChecker) Name() string {
+	return "module-check"
+}
+
+// Check determines if the modules specified with r.Modules have been loaded
+func (r ModuleChecker) Check(ctx context.Context, reporter health.Reporter) {
+	err := r.check(ctx, reporter)
+	if err != nil {
+		reporter.Add(NewProbeFromErr(r.Name(), "", trace.Wrap(err)))
+		return
+	}
+	reporter.Add(&pb.Probe{Checker: r.Name(), Status: pb.Probe_Running})
+}
+
+func (r ModuleChecker) check(ctx context.Context, reporter health.Reporter) error {
+	modules, err := r.moduleGetter.Modules()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var errors []error
+	for _, module := range r.Modules {
+		if !modules.HasLoaded(module) {
+			errors = append(errors, trace.NotFound("module %q not loaded", module))
+		}
+	}
+
+	if len(errors) == 1 {
+		return trace.Wrap(errors[0])
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// ModuleChecker checks if the specified set of kernel modules ara loaded
+type ModuleChecker struct {
+	// Modules lists required kernel modules
+	Modules []string
+	moduleGetter
+}
+
+// LoadModules loads list of kernel modules from /proc/modules
+func LoadModules() (Modules, error) {
+	f, err := os.Open("/proc/modules")
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+	defer f.Close()
+
+	modules, err := LoadModulesFrom(f)
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	return modules, nil
+}
+
+// LoadModulesFrom loads list of kernel modules from the specified reader
+func LoadModulesFrom(r io.Reader) (modules Modules, err error) {
+	s := bufio.NewScanner(r)
+
+	for s.Scan() {
+		line := s.Text()
+		module, err := parseModule(line)
+		if err != nil {
+			log.Warnf("failed to parse module, skip line %q", line)
+			continue
+		}
+		modules = append(modules, *module)
+	}
+
+	if s.Err() != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	return modules, nil
+}
+
+// HasLoaded determines whether module names are loaded.
+// names should not have duplicates
+func (r Modules) HasLoaded(names ...string) bool {
+L:
+	for _, module := range r {
```

Comment:
if Modules were a map this check will be trivial, also why re-implement loop here? WasLoaded(name) will work as expected and users can loop around when necessary to check the list. This function is doing too much IMHO